lib/Sema/AvailableFactsAnalysis.cpp

```diff
@@ -0,0 +1,499 @@
+    switch (BO->getOpcode()) {
+      case BinaryOperatorKind::BO_LE:
+      case BinaryOperatorKind::BO_LT:
+        ISet.insert(Comparison(BO->getLHS(), BO->getRHS()));
+        break;
+      case BinaryOperatorKind::BO_GE:
+      case BinaryOperatorKind::BO_GT:
+        ISet.insert(Comparison(BO->getRHS(), BO->getLHS()));
+        break;
+      case BinaryOperatorKind::BO_EQ:
+        ISet.insert(Comparison(BO->getRHS(), BO->getLHS()));
+        ISet.insert(Comparison(BO->getLHS(), BO->getRHS()));
+        break;
+      case BinaryOperatorKind::BO_LAnd:
+        ExtractComparisons(BO->getRHS(), ISet);
+        ExtractComparisons(BO->getLHS(), ISet);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+// This function computes a list of negated comparisons from `E`.
+// - If `E` is a simple direct comparison expression `A op B`, then the negated
+//   comparison can be created if `op` is one of LE, LT, GE, GT, or NE.
+// - If `E` has the form `A || B`, negated comparisons can be created for A and B.
+// Note that we do not include comparisons whose expressions involve
+// function calls or references to volatile variables.
+//
+// Some examples:
+// - `E` has the form A < B: add (B, A) to `ISet`.
+// - `E` has the form `E1 || E2`: ExtractNegatedComparisons in E1 and E2 and add
+//   them to `ISet`.
+void AvailableFactsAnalysis::ExtractNegatedComparisons(const Expr *E, ComparisonSet &ISet) {
+  if (IsVolatile(E) || ContainsCallExpr(E))
+    return;
+  if (const BinaryOperator *BO = dyn_cast<BinaryOperator>(E->IgnoreParens()))
+    switch (BO->getOpcode()) {
+      case BinaryOperatorKind::BO_LE:
+      case BinaryOperatorKind::BO_LT:
+        ISet.insert(Comparison(BO->getRHS(), BO->getLHS()));
+        break;
+      case BinaryOperatorKind::BO_GE:
+      case BinaryOperatorKind::BO_GT:
+        ISet.insert(Comparison(BO->getLHS(), BO->getRHS()));
+        break;
+      case BinaryOperatorKind::BO_NE:
+        ISet.insert(Comparison(BO->getRHS(), BO->getLHS()));
+        ISet.insert(Comparison(BO->getLHS(), BO->getRHS()));
+        break;
+      case BinaryOperatorKind::BO_LOr:
+        ExtractNegatedComparisons(BO->getRHS(), ISet);
+        ExtractNegatedComparisons(BO->getLHS(), ISet);
+        break;
+      default:
+        break;
+    }
+}
+
+void AvailableFactsAnalysis::CollectExpressions(const Stmt *St, std::set<const Expr *> &AllExprs) {
+  if (!St)
+    return;
+  if (const Expr *E = dyn_cast<Expr>(St))
+    AllExprs.insert(E);
+  for (auto I = St->child_begin(); I != St->child_end(); ++I)
+    CollectExpressions(*I, AllExprs);
+}
+
+// This function collects the defined variables in statement `St`.
+// We assume a variable is defined if it appears in the lhs of an assignment:
+// 1. increment or decrement operator (a++)
+// 2. assignment operator (a += 1, a = 2)
+void AvailableFactsAnalysis::CollectDefinedVars(const Stmt *St, std::set<const VarDecl *> &DefinedVars) {
```

Comment:
This function could miss some definitions of variables.  In the unary operator case, the expression being incremented/decremented could be parenthesized.    In both cases, the LValueBitCast could also be parenthesized.  I think you need a loop that ignores parentheses and lvalue bit casts until nothing changes.  This should be factored into a separate function.